pkg/hosts/hosts.go

```diff
@@ -88,57 +89,85 @@ func (h *Hosts) Add(ipRaw string, hosts []string) error {
 		return err
 	}
 
-	line, err := h.findIP(start, end, ip)
+	lines, err := h.findIP(start, end, ip)
 	if err != nil {
 		return err
 	}
 
 	// no host record, need to create new host line
-	if line == nil {
-		hfl := libhosty.HostsFileLine{
-			Type:      libhosty.LineTypeAddress,
-			Address:   ip,
-			Hostnames: hostEntries,
+	if lines == nil {
+		if len(hostEntries) > maxHostsInLine {
+			multiLineHosts := ChunkSlice(hostEntries, maxHostsInLine)
+			for _, hosts := range multiLineHosts {
+				h.createAndAddHostsLine(ip, hosts, start)
+			}
+		} else {
+			h.createAndAddHostsLine(ip, hostEntries, start)
 		}
 
-		// inserts to hosts
-		newHosts := make([]libhosty.HostsFileLine, 0)
-		newHosts = append(newHosts, h.File.HostsFileLines[:start+1]...)
-		newHosts = append(newHosts, hfl)
-		newLineNum := len(newHosts) - 1
-		h.File.Lock()
-		newHosts = append(newHosts, h.File.HostsFileLines[start+1:]...)
-		h.File.HostsFileLines = newHosts
-
-		// generate raw version of the line
-		hfl.Raw = h.File.RenderHostsFileLine(newLineNum)
-		h.File.Unlock()
-
 	} else {
 		var hostToAdd []string
 		for _, hostName := range hostEntries {
 			// check that new host not present in this line
 			contains := false
-			for _, lineHost := range line.Hostnames {
-				if hostName == lineHost {
-					contains = true
-					break
-				}
-			}
+
+			lineNum, _ := h.File.GetHostsFileLineByHostname(hostName)
+
+			// check that line inside crc region
+			contains = lineNum > start && lineNum < end
+
 			// add only new hosts
 			if !contains {
 				hostToAdd = append(hostToAdd, hostName)
 			}
 		}
-		h.File.Lock()
-		line.Hostnames = append(line.Hostnames, hostToAdd...)
-		h.File.Unlock()
+		for _, line := range lines {
+			if len(hostToAdd)+len(line.Hostnames) > maxHostsInLine {
+
+				fittingNumOfRecords := maxHostsInLine - len(line.Hostnames)
+				if fittingNumOfRecords > len(hostToAdd) {
+					fittingNumOfRecords = len(hostToAdd)
+				}
+
+				hostsForExistingLine := hostToAdd[0:fittingNumOfRecords]
+				h.File.Lock()
+				line.Hostnames = append(line.Hostnames, hostsForExistingLine...)
+				h.File.Unlock()
+				hostToAdd = hostToAdd[fittingNumOfRecords:]
+
+			} else {
+				h.File.Lock()
+				line.Hostnames = append(line.Hostnames, hostToAdd...)
+				h.File.Unlock()
+			}
```

Comment:
I'd try to move most of this logic to `ChunkSlice`, or a an additional (but similar) `func ChunkSlices(chunkSize int, slices ..[]string) [][]string`

Reply:
2 corner cases which fail below. 1st one is existing line with 9 entries, adding a 10th one. 2nd one is with an incorrect existing line with 10 entries, and adding a 11th entry.

```
+func TestAddMoreThan9HostsWithFullLine(t *testing.T) {
+       dir, err := os.MkdirTemp("", "hosts")
+       assert.NoError(t, err)
+       defer os.RemoveAll(dir)
+
+       hostsFile := filepath.Join(dir, "hosts")
+       assert.NoError(t, os.WriteFile(hostsFile, []byte(hostsTemplate+eol()+crcSection("127.0.0.1        entry1  entry2 entry3 entry4 entry5 entry6 entry7 entry8 entry9")+eol()), 0600))
+
+       host := hosts(t, hostsFile)
+
+       assert.NoError(t, host.Add("127.0.0.1", []string{"entry10"}))
+
+       content, err := os.ReadFile(hostsFile)
+       assert.NoError(t, err)
+       assert.Equal(t, hostsTemplate+eol()+crcSection("127.0.0.1        entry1 entry2 entry3 entry4 entry5 entry6 entry7 entry8 entry9", "127.0.0.1        entry10")+eol(), string(content))
+}
+
+func TestAddMoreThan9HostsWithOverfullLine(t *testing.T) {
+       dir, err := os.MkdirTemp("", "hosts")
+       assert.NoError(t, err)
+       defer os.RemoveAll(dir)
+
+       hostsFile := filepath.Join(dir, "hosts")
+       assert.NoError(t, os.WriteFile(hostsFile, []byte(hostsTemplate+eol()+crcSection("127.0.0.1        entry1  entry2 entry3 entry4 entry5 entry6 entry7 entry8 entry9 entry10")+eol()), 0600))
+
+       host := hosts(t, hostsFile)
+
+       assert.NoError(t, host.Add("127.0.0.1", []string{"entry11"}))
+
+       content, err := os.ReadFile(hostsFile)
+       assert.NoError(t, err)
+       assert.Equal(t, hostsTemplate+eol()+crcSection("127.0.0.1        entry1 entry2 entry3 entry4 entry5 entry6 entry7 entry8 entry9", "127.0.0.1        entry10 entry11")+eol(), string(content))
+}
+
```